### PR TITLE
Make VariableVersion refcounting thread-safe

### PR DESCRIPTION
**Pull request by colesbury**
I've also made the version counter and the "live" reference count
atomics.

Note that it's not safe to set the version counter (operator=) from
multiple threads, because shared_ptr assignment isn't thread safe.
Currently, the only call sites to these functions are on newly created
variables before they can be accessed from other threads.

See #3111

### CODE REVIEW

**Comment by ezyang** [comment]
Someday, when we have perf infra, it would be good to know what changes like this cost :)

File: tools/autograd/templates/VariableType.cpp [hunk 1]
```diff
@@ -200,16 +200,18 @@ static void wrap_output(std::tuple<Variable, Variable, Variable>& t, FunctionFla
 }
 
 static void increment_version(const Tensor & t) {
-  auto pImpl = static_cast<VariableImpl*>(t.get());
-  (*pImpl->version_counter)++;
+  // accept a const Tensor & because some functions (e.g. relu) mark inputs as
+  // const even though they may be modified in-place.
```
Comment (ezyang) [inline tools/autograd/templates/VariableType.cpp]: The const correctness story of ATen is a disaster (stemming from the disaster that is C++ const correctness, see https://github.com/zdevito/ATen/issues/27) but the consistent approach would be to make `version_counter()` in `Variable` self-const. I think you'll find that will compile.

File: torch/csrc/autograd/saved_variable.h [hunk 2]
```diff
@@ -37,7 +37,7 @@ struct SavedVariable {
   bool has_grad_fn;
   std::shared_ptr<Function> grad_fn;
   std::weak_ptr<Function> grad_accumulator;
-  std::unique_ptr<VariableVersion> version;
+  SavedVersion version;
```
Comment (ezyang) [inline torch/csrc/autograd/saved_variable.h]: Does this fix #3006?

**Comment by colesbury** [comment]
Yeah, I would really like automated benchmarking. My current test is basically (in ipython3):

```
import torch; from torch.autograd import Variable; v = Variable(torch.randn(1));
%timeit var.add(var)
```

This did not noticeably change.

With the PR, we save on one small allocation/free (~30ns) and add a few atomic ops (~5ns each).

File: torch/csrc/autograd/saved_variable.h (hunk 2 above)
Comment (colesbury) [reply torch/csrc/autograd/saved_variable.h]: Yeah

File: torch/csrc/autograd/variable_version.h [hunk 3]
```diff
@@ -4,52 +4,95 @@
...
-  void join_with(VariableVersion &other) {
-    if (this == &other) {
-      return;
-    }
-    cleanup();
+  // creates a saved reference with the current version and the counter
+  inline SavedVersion save() const;
+
+  // Uses another variable's version counter. Used for variables which share storages
+  // NOTE: not thread-safe to call this from multiple threads without synchronization
```
Comment (ezyang) [inline torch/csrc/autograd/variable_version.h]: Just to make sure I understand, this is because the `shared_ptr` is not atomic, right?

**Comment by ezyang** [comment]
I'm not convinced the new code is completely race free, but this is definitely an improvement and we should take it.

**Review by ezyang — APPROVED** [review]
LGTM.

**Review by apaszke — APPROVED** [review]
LGTM

File: torch/csrc/autograd/variable_version.h (hunk 3 above)
Comment (colesbury) [reply torch/csrc/autograd/variable_version.h]: Yes, that's all

**Pull request merged**